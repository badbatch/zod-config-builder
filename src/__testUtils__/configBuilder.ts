import { kebabCase } from 'lodash-es';
import { createConfigBuilder } from '../createConfigBuilder.ts';
import {
  type ConfigType,
  type PageType,
  type RouteType,
  type SectionType,
  configSchema,
  pageSchema,
  routeSchema,
  sectionSchema,
} from './schema.ts';

const configBuilder = createConfigBuilder<ConfigType>(configSchema);
const routeBuilder = createConfigBuilder<RouteType>(routeSchema, { path: ({ page }) => kebabCase(page) });
const pageBuilder = createConfigBuilder<PageType>(pageSchema);
const sectionBuilder = createConfigBuilder<SectionType>(sectionSchema);
const subsectionBuilder = sectionBuilder.fork();

configBuilder
  .countryCode('GB')
  .countryName('United Kingdom')
  .distanceUnit('km')
  .languageCodes(['en'])
  .locales(({ countryCode, languageCodes }) =>
    languageCodes?.length && countryCode ? languageCodes.map(code => `${code}_${countryCode}`) : []
  )
  .name('alpha')
  .pages({
    contactDetails: pageBuilder
      .name('contactDetails')
      .sections([
        sectionBuilder.name('header').flush(),
        sectionBuilder
          .name('body')
          .sections([subsectionBuilder.name('main').flush(), subsectionBuilder.name('sidebar').flush()])
          .flush(),
        sectionBuilder.name('footer').flush(),
      ])
      .flush(),
    personalDetails: pageBuilder
      .name('personalDetails')
      .sections([
        sectionBuilder.name('header').flush(),
        sectionBuilder
          .name('body')
          .sections([subsectionBuilder.name('main').flush(), subsectionBuilder.name('sidebar').flush()])
          .flush(),
        sectionBuilder.name('footer').flush(),
      ])
      .flush(),
  })
  .routes([routeBuilder.page('personalDetails').flush(), routeBuilder.page('contactDetails').flush()])
  .timeouts({ apollo: 10_000 })
  .timezone('Europe/London');

// eslint-disable-next-line import/no-default-export
export default configBuilder;
