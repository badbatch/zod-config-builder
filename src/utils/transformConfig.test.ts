import configBuilder from '../__testUtils__/configBuilder.ts';

describe('transformConfig', () => {
  describe('when no handlers are passed in', () => {
    it('should return the correct config', async () => {
      const { transformConfig } = await import('./transformConfig.ts');

      await expect(transformConfig(configBuilder.$values())).resolves.toMatchInlineSnapshot(`
        {
          "countryCode": "GB",
          "countryName": null,
          "distanceUnit": "km",
          "enabled": true,
          "languageCodes": [
            "en",
          ],
          "locales": [
            "en_GB",
          ],
          "name": undefined,
          "pages": {
            "contactDetails": {
              "name": "contactDetails",
              "path": undefined,
              "queryParams": undefined,
              "sections": [
                {
                  "name": "header",
                  "sections": undefined,
                },
                {
                  "name": "body",
                  "sections": [
                    {
                      "name": "main",
                      "sections": undefined,
                    },
                    {
                      "name": "sidebar",
                    },
                  ],
                },
                {
                  "name": "footer",
                },
              ],
            },
            "personalDetails": {
              "name": "personalDetails",
              "sections": [
                {
                  "name": "header",
                },
                {
                  "name": "body",
                  "sections": [
                    {
                      "name": "main",
                    },
                    {
                      "name": "sidebar",
                    },
                  ],
                },
                {
                  "name": "footer",
                },
              ],
            },
          },
          "routes": [
            {
              "aliases": undefined,
              "page": "personalDetails",
              "path": "personal-details",
              "routes": undefined,
            },
            {
              "page": "contactDetails",
              "path": "contact-details",
            },
          ],
          "timeouts": {
            "apollo": 10000,
          },
          "timezone": "Europe/London",
        }
      `);
    });
  });
});

describe('transformConfigSync', () => {
  describe('when no handlers are passed in', () => {
    it('should return the correct config', async () => {
      const { transformConfigSync } = await import('./transformConfig.ts');

      expect(transformConfigSync(configBuilder.$values())).toMatchInlineSnapshot(`
        {
          "countryCode": "GB",
          "countryName": null,
          "distanceUnit": "km",
          "enabled": true,
          "languageCodes": [
            "en",
          ],
          "locales": [
            "en_GB",
          ],
          "name": undefined,
          "pages": {
            "contactDetails": {
              "name": "contactDetails",
              "path": undefined,
              "queryParams": undefined,
              "sections": [
                {
                  "name": "header",
                  "sections": undefined,
                },
                {
                  "name": "body",
                  "sections": [
                    {
                      "name": "main",
                      "sections": undefined,
                    },
                    {
                      "name": "sidebar",
                    },
                  ],
                },
                {
                  "name": "footer",
                },
              ],
            },
            "personalDetails": {
              "name": "personalDetails",
              "sections": [
                {
                  "name": "header",
                },
                {
                  "name": "body",
                  "sections": [
                    {
                      "name": "main",
                    },
                    {
                      "name": "sidebar",
                    },
                  ],
                },
                {
                  "name": "footer",
                },
              ],
            },
          },
          "routes": [
            {
              "aliases": undefined,
              "page": "personalDetails",
              "path": "personal-details",
              "routes": undefined,
            },
            {
              "page": "contactDetails",
              "path": "contact-details",
            },
          ],
          "timeouts": {
            "apollo": 10000,
          },
          "timezone": "Europe/London",
        }
      `);
    });
  });
});
