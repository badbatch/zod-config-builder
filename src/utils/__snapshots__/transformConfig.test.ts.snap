// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformConfig when no handlers are passed in should return the correct config 1`] = `
{
  "countryCode": "GB",
  "countryName": "United Kingdom",
  "distanceUnit": "km",
  "languageCodes": [
    "en",
  ],
  "locales": [
    "en_GB",
  ],
  "name": "alpha",
  "pages": {
    "contactDetails": {
      "name": "contactDetails",
      "sections": [
        {
          "name": "header",
        },
        {
          "name": "body",
          "sections": [
            {
              "name": "main",
            },
            {
              "name": "sidebar",
            },
          ],
        },
        {
          "name": "footer",
        },
      ],
    },
    "personalDetails": {
      "name": "personalDetails",
      "sections": [
        {
          "name": "header",
        },
        {
          "name": "body",
          "sections": [
            {
              "name": "main",
            },
            {
              "name": "sidebar",
            },
          ],
        },
        {
          "name": "footer",
        },
      ],
    },
  },
  "routes": [
    {
      "page": "personalDetails",
      "path": "personal-details",
    },
    {
      "page": "contactDetails",
      "path": "contact-details",
    },
  ],
  "timeouts": {
    "apollo": 10000,
  },
  "timezone": "Europe/London",
}
`;

exports[`transformConfigSync when no handlers are passed in should return the correct config 1`] = `
{
  "countryCode": "GB",
  "countryName": "United Kingdom",
  "distanceUnit": "km",
  "languageCodes": [
    "en",
  ],
  "locales": [
    "en_GB",
  ],
  "name": "alpha",
  "pages": {
    "contactDetails": {
      "name": "contactDetails",
      "sections": [
        {
          "name": "header",
        },
        {
          "name": "body",
          "sections": [
            {
              "name": "main",
            },
            {
              "name": "sidebar",
            },
          ],
        },
        {
          "name": "footer",
        },
      ],
    },
    "personalDetails": {
      "name": "personalDetails",
      "sections": [
        {
          "name": "header",
        },
        {
          "name": "body",
          "sections": [
            {
              "name": "main",
            },
            {
              "name": "sidebar",
            },
          ],
        },
        {
          "name": "footer",
        },
      ],
    },
  },
  "routes": [
    {
      "page": "personalDetails",
      "path": "personal-details",
    },
    {
      "page": "contactDetails",
      "path": "contact-details",
    },
  ],
  "timeouts": {
    "apollo": 10000,
  },
  "timezone": "Europe/London",
}
`;
